## Web APIs - The Alternative to Scraping

Web scraping doesn't always work. Many modern websites are built in a way that makes it particularly hard to scrape. They may do one of the following:

- Incorporate "infinite scrolling", so that the page that is loaded is not the same as the content you want to scrape
- Block scraping through bot detection
- Have a website sufficiently complex that scraping is too hard to implement

Why would websites actively discourage or even make scraping impossible? For the most part, we can imagine websites as a storefront. The website is designed to provide some good or service in exchange for either money (through purchases) or in exchange for the consumption of advertisements or other forms of compensation. Although these "storefronts" are online, they are not costless.

Every time a website is requested, the server must send that information to the person/computer/bot that has requested it. This might be very cheap for a single request, but these costs add up when hundreds, thousands, or even millions of requests must be processed. When we scrape a website, we are in essence trying to collect a good (data) without compensating the group who maintains the website. For this reason, many websites will seek to block web scraping whenever they detect it, and will include a statement to this effect in their terms of service.

In cases where scraping is not feasible for one reason or another, we will need an alternative way to collect data. Web APIs will be our key to data collection in these situations.


## What are web APIs?

The acronym **API** stands for **Application Programming Interface**, and is the name for the way in which a program or user may communicate with another program. Using an API enables the creation of explicit mechanisms for input and output of data.

A **web API** is simply the API that allows users and programs to interact with a website or web service, so that we can automate or streamline online interactions. APIs are used for the following (and much more!):

- 3rd party social media apps
- Map interfaces on business websites
- Data collection


### Comparing APIs to Scraping

When using a web API, our process is very different than when we scrape. As discussed earlier, scraping typically occurs at the expense of the service provider, making many websites averse to allowing scraping to take place. On the other hand, when using an API, there is an explicit agreement between the provider of data and the consumer. In many cases, APIs cost money to use, and are designed to compensate the service provider in a transaction reflecting the fact that the consumer no longer seeks to buy a good/service on the website or view advertisements. Instead, the data being transferred is the product to be consumed.

While we will use a free and open API to learn during this class, bear in mind that many (if not most) APIs will require the user to register before making requests, and may also require payment. This arrangement allows both users and providers to benefit from the exchange of data.


## Using a web API

To learn about using web APIs, we will experiment with the free location API provided by [zippopotam.us](http://www.zippopotam.us/). We will learn the basics of API use with the simple examples that can be implemented through the Zippopotamus API, and hopefully make some fun graphs while we are at it!

Most web APIs are accessed through the use of a URL, just like any other website. We then use the various extensions of the URL to make a request for an API call, rather than to request a specific website. Let's give this a try with the `requests` library, and see what we get back as we request various URLs from Zippopotamus:

In [2]:
import requests

requests.get("https://api.zippopotam.us/us/68022").text

requests.get('https://api.zippopotam.us/us/ma/belmont').text

'{"country abbreviation": "US", "places": [{"place name": "Belmont", "longitude": "-71.4594", "post code": "02178", "latitude": "42.4464"}, {"place name": "Belmont", "longitude": "-71.2044", "post code": "02478", "latitude": "42.4128"}], "country": "United States", "place name": "Belmont", "state": "Massachusetts", "state abbreviation": "MA"}'

    '{"post code": "68022", "country": "United States", "country abbreviation": "US", "places": [{"place name": "Elkhorn", "longitude": "-96.2431", "state": "Nebraska", "state abbreviation": "NE", "latitude": "41.2756"}]}'

Cool! We got some random stuff when we requested a random page from Zippopotamus... What is this? Let's break down our URL to learn more about what is going on:

> Add blockquote

> Add blockquote





 - `https://api.zippopotam.us` is the URL that we will always call when making requests to or from Zippopotamus
 - `/us` tells Zippopotamus that the location we are referencing is in the United States
 - `/68022` is the postal code for which we want to extract information from Zippopotamus

The data that we receive back from our request is information about the postal code that we want to examine. Before we start learning how to work with our results, let's try another API call:

requests.get("https://api.zippopotam.us/us/wa/redmond").text

    '{"country abbreviation": "US", "places": [{"place name": "Redmond", "longitude": "-122.1232", "post code": "98052", "latitude": "47.6718"}, {"place name": "Redmond", "longitude": "-122.0386", "post code": "98053", "latitude": "47.6462"}, {"place name": "Redmond", "longitude": "-121.8034", "post code": "98073", "latitude": "47.4323"}], "country": "United States", "place name": "Redmond", "state": "Washington", "state abbreviation": "WA"}'



In this new request, we have the following:

- `https://api.zippopotam.us` is the URL that we will always call when making requests to or from Zippopotamus
- `/us` tells Zippopotamus that the location we are referencing is in the United States
- `/wa` tells Zippopotamus that the location is in Washington state
- `/redmond` is the city for which we would like to extract information

As might be evident from the differences between the two requests we have made, many APIs are able to access different kinds of information depending on the different requests that we make. We can find information on the city in which a postal code is located, or we can find postal codes within a city of interest! This will allow us to make requests that will generate a map of places that I have lived during my life. Let's try it out.

In [3]:
livedThere = [
    {"country": "/us", "postal_code": "/92056"},
    {"country": "/us", "postal_code": "/98052"},
    {"country": "/us", "postal_code": "/84602"},
    {"country": "/br", "postal_code": "/96225-000"},
    {"country": "/br", "postal_code": "/96600-000"},
    {"country": "/us", "postal_code": "/99163"},
    {"country": "/us", "postal_code": "/68022"},
    {"country": "/us", "postal_code": "/84102"},
]

The cell above contains a list of dictionaries. I will iterate over this list to create my API calls:

In [4]:
for i in livedThere:
    call = "https://api.zippopotam.us{0}{1}".format(i["country"],i["postal_code"])
    print(call)

https://api.zippopotam.us/us/92056
https://api.zippopotam.us/us/98052
https://api.zippopotam.us/us/84602
https://api.zippopotam.us/br/96225-000
https://api.zippopotam.us/br/96600-000
https://api.zippopotam.us/us/99163
https://api.zippopotam.us/us/68022
https://api.zippopotam.us/us/84102


    https://api.zippopotam.us/us/92056
    https://api.zippopotam.us/us/98052
    https://api.zippopotam.us/us/84602
    https://api.zippopotam.us/br/96225-000
    https://api.zippopotam.us/br/96600-000
    https://api.zippopotam.us/us/99163
    https://api.zippopotam.us/us/68022
    https://api.zippopotam.us/us/84102


The above code combines my country codes for each location with the postal code and the base-URL to provide my API call. Now it is time to write some code to retrieve each request and store the results in a new list:

In [17]:
locales = []

for i in livedThere:
    call = "https://api.zippopotam.us{0}{1}".format(i["country"],i["postal_code"])
    locales.append(requests.get(call).text)

locales

['{"post code": "92056", "country": "United States", "country abbreviation": "US", "places": [{"place name": "Oceanside", "longitude": "-117.2831", "state": "California", "state abbreviation": "CA", "latitude": "33.1968"}]}',
 '{"post code": "98052", "country": "United States", "country abbreviation": "US", "places": [{"place name": "Redmond", "longitude": "-122.1232", "state": "Washington", "state abbreviation": "WA", "latitude": "47.6718"}]}',
 '{"post code": "84602", "country": "United States", "country abbreviation": "US", "places": [{"place name": "Provo", "longitude": "-111.7325", "state": "Utah", "state abbreviation": "UT", "latitude": "40.3563"}]}',
 '{"post code": "96225-000", "country": "Brazil", "country abbreviation": "BR", "places": [{"place name": "S\\u00e3o Jos\\u00e9 do Norte", "longitude": "-51.762", "state": "Rio Grande do Sul", "state abbreviation": "23", "latitude": "-31.8009"}]}',
 '{"post code": "96600-000", "country": "Brazil", "country abbreviation": "BR", "plac

    ['{"post code": "92056", "country": "United States", "country abbreviation": "US", "places": [{"place name": "Oceanside", "longitude": "-117.2831", "state": "California", "state abbreviation": "CA", "latitude": "33.1968"}]}',
     '{"post code": "98052", "country": "United States", "country abbreviation": "US", "places": [{"place name": "Redmond", "longitude": "-122.1232", "state": "Washington", "state abbreviation": "WA", "latitude": "47.6718"}]}',
     '{"post code": "84602", "country": "United States", "country abbreviation": "US", "places": [{"place name": "Provo", "longitude": "-111.7325", "state": "Utah", "state abbreviation": "UT", "latitude": "40.3563"}]}',
     '{"post code": "96225-000", "country": "Brazil", "country abbreviation": "BR", "places": [{"place name": "S\\u00e3o Jos\\u00e9 do Norte", "longitude": "-51.762", "state": "Rio Grande do Sul", "state abbreviation": "23", "latitude": "-31.8009"}]}',
     '{"post code": "96600-000", "country": "Brazil", "country abbreviation": "BR", "places": [{"place name": "Cangu\\u00e7u", "longitude": "-52.6682", "state": "Rio Grande do Sul", "state abbreviation": "23", "latitude": "-31.2117"}]}',
     '{"post code": "99163", "country": "United States", "country abbreviation": "US", "places": [{"place name": "Pullman", "longitude": "-117.1729", "state": "Washington", "state abbreviation": "WA", "latitude": "46.7352"}]}',
     '{"post code": "68022", "country": "United States", "country abbreviation": "US", "places": [{"place name": "Elkhorn", "longitude": "-96.2431", "state": "Nebraska", "state abbreviation": "NE", "latitude": "41.2756"}]}',
     '{"post code": "84102", "country": "United States", "country abbreviation": "US", "places": [{"place name": "Salt Lake City", "longitude": "-111.8627", "state": "Utah", "state abbreviation": "UT", "latitude": "40.76"}]}']



Awesome! We have the results! Except, they are really garbled....now we need to take these results and clean them up into a useable format. Let's figure out how.

## Processing `requests` results

Each response from a web API will typically be returned in a **JSON** object. **JSON** stands for **Java Script Object Notation**, and is the standard format for transmitting information on the internet. While this may sound confusing, JSON is actually very easy for us to handle in Python, because it is structured EXACTLY like a Python dictionary. All we need is a way to take a JSON object and translate it into an official `dict` object.

Lucky for us, the `json` library (built right into Python itself!) will make this an easy exercise.

In [6]:
import json

locales = [json.loads(i) for i in locales]

locales

[{'post code': '92056',
  'country': 'United States',
  'country abbreviation': 'US',
  'places': [{'place name': 'Oceanside',
    'longitude': '-117.2831',
    'state': 'California',
    'state abbreviation': 'CA',
    'latitude': '33.1968'}]},
 {'post code': '98052',
  'country': 'United States',
  'country abbreviation': 'US',
  'places': [{'place name': 'Redmond',
    'longitude': '-122.1232',
    'state': 'Washington',
    'state abbreviation': 'WA',
    'latitude': '47.6718'}]},
 {'post code': '84602',
  'country': 'United States',
  'country abbreviation': 'US',
  'places': [{'place name': 'Provo',
    'longitude': '-111.7325',
    'state': 'Utah',
    'state abbreviation': 'UT',
    'latitude': '40.3563'}]},
 {'post code': '96225-000',
  'country': 'Brazil',
  'country abbreviation': 'BR',
  'places': [{'place name': 'São José do Norte',
    'longitude': '-51.762',
    'state': 'Rio Grande do Sul',
    'state abbreviation': '23',
    'latitude': '-31.8009'}]},
 {'post code': '9

    [{'post code': '92056',
      'country': 'United States',
      'country abbreviation': 'US',
      'places': [{'place name': 'Oceanside',
        'longitude': '-117.2831',
        'state': 'California',
        'state abbreviation': 'CA',
        'latitude': '33.1968'}]},
     {'post code': '98052',
      'country': 'United States',
      'country abbreviation': 'US',
      'places': [{'place name': 'Redmond',
        'longitude': '-122.1232',
        'state': 'Washington',
        'state abbreviation': 'WA',
        'latitude': '47.6718'}]},
     {'post code': '84602',
      'country': 'United States',
      'country abbreviation': 'US',
      'places': [{'place name': 'Provo',
        'longitude': '-111.7325',
        'state': 'Utah',
        'state abbreviation': 'UT',
        'latitude': '40.3563'}]},
     {'post code': '96225-000',
      'country': 'Brazil',
      'country abbreviation': 'BR',
      'places': [{'place name': 'São José do Norte',
        'longitude': '-51.762',
        'state': 'Rio Grande do Sul',
        'state abbreviation': '23',
        'latitude': '-31.8009'}]},
     {'post code': '96600-000',
      'country': 'Brazil',
      'country abbreviation': 'BR',
      'places': [{'place name': 'Canguçu',
        'longitude': '-52.6682',
        'state': 'Rio Grande do Sul',
        'state abbreviation': '23',
        'latitude': '-31.2117'}]},
     {'post code': '99163',
      'country': 'United States',
      'country abbreviation': 'US',
      'places': [{'place name': 'Pullman',
        'longitude': '-117.1729',
        'state': 'Washington',
        'state abbreviation': 'WA',
        'latitude': '46.7352'}]},
     {'post code': '68022',
      'country': 'United States',
      'country abbreviation': 'US',
      'places': [{'place name': 'Elkhorn',
        'longitude': '-96.2431',
        'state': 'Nebraska',
        'state abbreviation': 'NE',
        'latitude': '41.2756'}]},
     {'post code': '84102',
      'country': 'United States',
      'country abbreviation': 'US',
      'places': [{'place name': 'Salt Lake City',
        'longitude': '-111.8627',
        'state': 'Utah',
        'state abbreviation': 'UT',
        'latitude': '40.76'}]}]



And we are done making our data into Python data! That was easy! Next, let's just take this list of dictionaries, and make it into a Data Frame. It turns out that we can write a little bit of code to make Data Frames out of our API results through `pandas`:

In [7]:
import pandas as pd

locales = pd.DataFrame(locales)

data = []
for i in locales.index:
    temp = pd.DataFrame(locales.loc[i, 'places'])
    temp['post_code'] = locales.loc[i, 'post code']
    temp['country'] = locales.loc[i, 'country']
    temp['country_code'] = locales.loc[i, 'country abbreviation']
    data.append(temp)

data = pd.concat(data, axis=0).reset_index(drop=True)

data

place name  longitude              state state abbreviation  \
0          Oceanside  -117.2831         California                 CA   
1            Redmond  -122.1232         Washington                 WA   
2              Provo  -111.7325               Utah                 UT   
3  São José do Norte    -51.762  Rio Grande do Sul                 23   
4            Canguçu   -52.6682  Rio Grande do Sul                 23   
5            Pullman  -117.1729         Washington                 WA   
6            Elkhorn   -96.2431           Nebraska                 NE   
7     Salt Lake City  -111.8627               Utah                 UT   

   latitude  post_code        country country_code  
0   33.1968      92056  United States           US  
1   47.6718      98052  United States           US  
2   40.3563      84602  United States           US  
3  -31.8009  96225-000         Brazil           BR  
4  -31.2117  96600-000         Brazil           BR  
5   46.7352      99163  United States           US  
6   41.2756      68022  United States           US  
7     40.76      84102  United States           US

<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>place name</th>
      <th>longitude</th>
      <th>state</th>
      <th>state abbreviation</th>
      <th>latitude</th>
      <th>post_code</th>
      <th>country</th>
      <th>country_code</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>Oceanside</td>
      <td>-117.2831</td>
      <td>California</td>
      <td>CA</td>
      <td>33.1968</td>
      <td>92056</td>
      <td>United States</td>
      <td>US</td>
    </tr>
    <tr>
      <th>1</th>
      <td>Redmond</td>
      <td>-122.1232</td>
      <td>Washington</td>
      <td>WA</td>
      <td>47.6718</td>
      <td>98052</td>
      <td>United States</td>
      <td>US</td>
    </tr>
    <tr>
      <th>2</th>
      <td>Provo</td>
      <td>-111.7325</td>
      <td>Utah</td>
      <td>UT</td>
      <td>40.3563</td>
      <td>84602</td>
      <td>United States</td>
      <td>US</td>
    </tr>
    <tr>
      <th>3</th>
      <td>São José do Norte</td>
      <td>-51.762</td>
      <td>Rio Grande do Sul</td>
      <td>23</td>
      <td>-31.8009</td>
      <td>96225-000</td>
      <td>Brazil</td>
      <td>BR</td>
    </tr>
    <tr>
      <th>4</th>
      <td>Canguçu</td>
      <td>-52.6682</td>
      <td>Rio Grande do Sul</td>
      <td>23</td>
      <td>-31.2117</td>
      <td>96600-000</td>
      <td>Brazil</td>
      <td>BR</td>
    </tr>
    <tr>
      <th>5</th>
      <td>Pullman</td>
      <td>-117.1729</td>
      <td>Washington</td>
      <td>WA</td>
      <td>46.7352</td>
      <td>99163</td>
      <td>United States</td>
      <td>US</td>
    </tr>
    <tr>
      <th>6</th>
      <td>Elkhorn</td>
      <td>-96.2431</td>
      <td>Nebraska</td>
      <td>NE</td>
      <td>41.2756</td>
      <td>68022</td>
      <td>United States</td>
      <td>US</td>
    </tr>
    <tr>
      <th>7</th>
      <td>Salt Lake City</td>
      <td>-111.8627</td>
      <td>Utah</td>
      <td>UT</td>
      <td>40.76</td>
      <td>84102</td>
      <td>United States</td>
      <td>US</td>
    </tr>
  </tbody>
</table>
</div>



Now that we have a nice, clean Data Frame containing the information on different locations, we can make a map out of it (we will cover mapping later in this course):

In [8]:
import plotly.express as px

px.scatter_geo(data, 'latitude', 'longitude',
               hover_data=['place name', 'state', 'country'],
               color='country',
               projection = 'natural earth',
               width = 900,
               height = 600,
               )

And there you have it! A map of places that I have lived!

![](newplot.png)

It really looks a lot less impressive on a world map...

APIs will allow us to do all sorts of amazing work. Some APIs that are awesome but require accounts are listed below:

- [Twitter Realtime Tweet Filter](https://developer.twitter.com/en/docs/tweets/filter-realtime/overview)
- [Tracker.gg](https://tracker.gg/)
- [Google Maps Distance Matrix](https://developers.google.com/maps/documentation/distance-matrix/intro)

With web scraping and APIs at our fingertips, our ability to gather and investigate data is limited only by our ability to come up with interesting research questions.

**Solve-it!**

In the cell commented with `#si-exercise` is a Data Frame called `capitals` containing capital cities in states across the United States. Use the [Zippopotam.us](https://api.zippopotam.us) API to gather latitude and longitude information on each city, and add columns named `lat` and `lon` containing the latitude and longitude data for each city, respectively. Once you have done that, map all of the cities using their latitude and longitude data on a map with `scope=usa`. Store the map as a variable called `fig`.

*Hints:*
1. If you use my code from the notes above, you can't use the projection that I used with `scope="usa"`. Get rid of the `projection` argument and replace it with the scope argument.
2. You just need to grab a single latitude and longitude for the cities. I recommend using the first one in the list.

You will be graded on the following:
- Created columns named `lat` and `lon` in the `capitals` Data Frame [1 point]
- `lat` contains correct latitude for the city [1 point]
- `lon` contains correct longitude for the city [1 point]
- Map (store as `fig`) displays capitals based on latitude and longitude values [1 point]

In [146]:
#si-exercise

import pandas as pd
import json
import plotly.express as px
import requests

capitals = pd.DataFrame([["Alabama","Montgomery","AL"],
["Alaska","Juneau","AK"],
["Arizona","Phoenix","AZ"],
["Arkansas","Little Rock","AR"],
["California","Sacramento","CA"],
["Colorado","Denver","CO"],
["Connecticut","Hartford","CT"],
["Delaware","Dover","DE"],
["Florida","Tallahassee","FL"],
["Georgia","Atlanta","GA"],
["Hawaii","Honolulu","HI"],
["Idaho","Boise","ID"],
["Illinois","Springfield","IL"],
["Indiana","Indianapolis","IN"],
["Iowa","Des Moines","IA"],
["Kansas","Topeka","KS"],
["Kentucky","Frankfort","KY"],
["Louisiana","Baton Rouge","LA"],
["Maine","Augusta","ME"],
["Maryland","Annapolis","MD"],
["Massachusetts","Boston","MA"],
["Michigan","Lansing","MI"],
["Minnesota","Saint Paul","MN"],
["Mississippi","Jackson","MS"],
["Missouri","Jefferson City","MO"],
["Montana","Helena","MT"],
["Nebraska","Lincoln","NE"],
["Nevada","Carson City","NV"],
["New Hampshire","Concord","NH"],
["New Jersey","Trenton","NJ"],
["New Mexico","Santa Fe","NM"],
["New York","Albany","NY"],
["North Carolina","Raleigh","NC"],
["North Dakota","Bismarck","ND"],
["Ohio","Columbus","OH"],
["Oklahoma","Oklahoma City","OK"],
["Oregon","Salem","OR"],
["Pennsylvania","Harrisburg","PA"],
["Rhode Island","Providence","RI"],
["South Carolina","Columbia","SC"],
["South Dakota","Pierre","SD"],
["Tennessee","Nashville","TN"],
["Texas","Austin","TX"],
["Utah","Salt Lake City","UT"],
["Vermont","Montpelier","VT"],
["Virginia","Richmond","VA"],
["Washington","Olympia","WA"],
["West Virginia","Charleston","WV"],
["Wisconsin","Madison","WI"],
["Wyoming","Cheyenne","WY"]], columns = ['state', 'capital', 'abbrev'])



###
info = []

for i in capitals.index:
    call = "https://api.zippopotam.us/us/{0}/{1}".format(capitals["abbrev"][i],capitals["capital"][i])
    info.append(requests.get(call).text)

info = [json.loads(i) for i in info]

dummy = []

for i in range(50):
  for j in range(len(info[i]['places'])):
    if info[i]['places'][j]['place name'] == capitals['capital'][i]:
      dummy.append(info[i]['places'][j])
      break

lon = []
lat = []

for i in range(50):
  lon.append(dummy[i]['longitude'])
  lat.append(dummy[i]['latitude'])

capitals.insert(0, "lon", lon, True)
capitals.insert(0, "lat", lat, True)

###

fig = px.scatter_geo(capitals, 'lat', 'lon', hover_data=['state', 'capital'], scope="usa")

In [148]:
capitals

lat        lon           state         capital abbrev
0    32.357   -86.2578         Alabama      Montgomery     AL
1   58.3628  -134.5294          Alaska          Juneau     AK
2    33.704  -112.3518         Arizona         Phoenix     AZ
3   34.7483   -92.2819        Arkansas     Little Rock     AR
4   38.3805  -121.5554      California      Sacramento     CA
5   39.7263  -104.8568        Colorado          Denver     CO
6   41.7801   -72.6771     Connecticut        Hartford     CT
7   39.1564   -75.4955        Delaware           Dover     DE
8   30.4286   -84.2593         Florida     Tallahassee     FL
9   33.8444    -84.474         Georgia         Atlanta     GA
10  24.8598  -168.0218          Hawaii        Honolulu     HI
11  43.6038  -116.2729           Idaho           Boise     ID
12     39.8   -89.6495        Illinois     Springfield     IL
13   39.775   -86.1093         Indiana    Indianapolis     IN
14  41.6727   -93.5722            Iowa      Des Moines     IA
15  38.9881   -95.7807          Kansas          Topeka     KS
16  38.2281   -84.8697        Kentucky       Frankfort     KY
17  30.4492   -91.1856       Louisiana     Baton Rouge     LA
18  44.3232   -69.7665           Maine         Augusta     ME
19  38.9898   -76.5501        Maryland       Annapolis     MD
20  42.3706    -71.027   Massachusetts          Boston     MA
21  42.7039   -84.6228        Michigan         Lansing     MI
22  44.9512   -93.0902       Minnesota      Saint Paul     MN
23  32.3205   -90.2076     Mississippi         Jackson     MS
24  38.5462   -92.1525        Missouri  Jefferson City     MO
25  46.6131  -112.0213         Montana          Helena     MT
26  40.8651   -96.8231        Nebraska         Lincoln     NE
27  39.1507  -119.7459          Nevada     Carson City     NV
28  43.2185   -71.5277   New Hampshire         Concord     NH
29  40.2805    -74.712      New Jersey         Trenton     NJ
30  35.6975  -105.9821      New Mexico        Santa Fe     NM
31  42.6149   -73.9708        New York          Albany     NY
32  35.7727   -78.6324  North Carolina         Raleigh     NC
33  46.8234  -100.7748    North Dakota        Bismarck     ND
34  40.0999   -83.0157            Ohio        Columbus     OH
35  35.3513   -97.4953        Oklahoma   Oklahoma City     OK
36   44.949   -123.004          Oregon           Salem     OR
37  40.2618   -76.8831    Pennsylvania      Harrisburg     PA
38  41.8228   -71.4145    Rhode Island      Providence     RI
39  34.0004   -81.0334  South Carolina        Columbia     SC
40  44.3695  -100.3211    South Dakota          Pierre     SD
41  36.1657   -86.7781       Tennessee       Nashville     TN
42  30.3264   -97.7713           Texas          Austin     TX
43  40.7559  -111.8967            Utah  Salt Lake City     UT
44  44.1991   -72.5596         Vermont      Montpelier     VT
45  37.5593   -77.4471        Virginia        Richmond     VA
46  47.0129  -122.8763      Washington         Olympia     WA
47   38.349   -81.6306   West Virginia      Charleston     WV
48  43.0696   -89.4239       Wisconsin         Madison     WI
49  41.1437  -104.7962         Wyoming        Cheyenne     WY

In [144]:

dummy = []

for i in range(50):
  for j in range(len(info[i]['places'])):
    if info[i]['places'][j]['place name'] == capitals['capital'][i]:
      dummy.append(info[i]['places'][j])
      break

lon = []
lat = []

for i in range(50):
  lon.append(dummy[i]['longitude'])
  lat.append(dummy[i]['latitude'])

capitals.insert(0, "lon", lon, True)
capitals.insert(0, "lat", lat, True)

capitals

lat        lon      lat        lon           state         capital  \
0    32.357   -86.2578   32.357   -86.2578         Alabama      Montgomery   
1   58.3628  -134.5294  58.3628  -134.5294          Alaska          Juneau   
2    33.704  -112.3518   33.704  -112.3518         Arizona         Phoenix   
3   34.7483   -92.2819  34.7483   -92.2819        Arkansas     Little Rock   
4   38.3805  -121.5554  38.3805  -121.5554      California      Sacramento   
5   39.7263  -104.8568  39.7263  -104.8568        Colorado          Denver   
6   41.7801   -72.6771  41.7801   -72.6771     Connecticut        Hartford   
7   39.1564   -75.4955  39.1564   -75.4955        Delaware           Dover   
8   30.4286   -84.2593  30.4286   -84.2593         Florida     Tallahassee   
9   33.8444    -84.474  33.8444    -84.474         Georgia         Atlanta   
10  24.8598  -168.0218  24.8598  -168.0218          Hawaii        Honolulu   
11  43.6038  -116.2729  43.6038  -116.2729           Idaho           Boise   
12     39.8   -89.6495     39.8   -89.6495        Illinois     Springfield   
13   39.775   -86.1093   39.775   -86.1093         Indiana    Indianapolis   
14  41.6727   -93.5722  41.6727   -93.5722            Iowa      Des Moines   
15  38.9881   -95.7807  38.9881   -95.7807          Kansas          Topeka   
16  38.2281   -84.8697  38.2281   -84.8697        Kentucky       Frankfort   
17  30.4492   -91.1856  30.4492   -91.1856       Louisiana     Baton Rouge   
18  44.3232   -69.7665  44.3232   -69.7665           Maine         Augusta   
19  38.9898   -76.5501  38.9898   -76.5501        Maryland       Annapolis   
20  42.3706    -71.027  42.3706    -71.027   Massachusetts          Boston   
21  42.7039   -84.6228  42.7039   -84.6228        Michigan         Lansing   
22  44.9512   -93.0902  44.9512   -93.0902       Minnesota      Saint Paul   
23  32.3205   -90.2076  32.3205   -90.2076     Mississippi         Jackson   
24  38.5462   -92.1525  38.5462   -92.1525        Missouri  Jefferson City   
25  46.6131  -112.0213  46.6131  -112.0213         Montana          Helena   
26  40.8651   -96.8231  40.8651   -96.8231        Nebraska         Lincoln   
27  39.1507  -119.7459  39.1507  -119.7459          Nevada     Carson City   
28  43.2185   -71.5277  43.2185   -71.5277   New Hampshire         Concord   
29  40.2805    -74.712  40.2805    -74.712      New Jersey         Trenton   
30  35.6975  -105.9821  35.6975  -105.9821      New Mexico        Santa Fe   
31  42.6149   -73.9708  42.6149   -73.9708        New York          Albany   
32  35.7727   -78.6324  35.7727   -78.6324  North Carolina         Raleigh   
33  46.8234  -100.7748  46.8234  -100.7748    North Dakota        Bismarck   
34  40.0999   -83.0157  40.0999   -83.0157            Ohio        Columbus   
35  35.3513   -97.4953  35.3513   -97.4953        Oklahoma   Oklahoma City   
36   44.949   -123.004   44.949   -123.004          Oregon           Salem   
37  40.2618   -76.8831  40.2618   -76.8831    Pennsylvania      Harrisburg   
38  41.8228   -71.4145  41.8228   -71.4145    Rhode Island      Providence   
39  34.0004   -81.0334  34.0004   -81.0334  South Carolina        Columbia   
40  44.3695  -100.3211  44.3695  -100.3211    South Dakota          Pierre   
41  36.1657   -86.7781  36.1657   -86.7781       Tennessee       Nashville   
42  30.3264   -97.7713  30.3264   -97.7713           Texas          Austin   
43  40.7559  -111.8967  40.7559  -111.8967            Utah  Salt Lake City   
44  44.1991   -72.5596  44.1991   -72.5596         Vermont      Montpelier   
45  37.5593   -77.4471  37.5593   -77.4471        Virginia        Richmond   
46  47.0129  -122.8763  47.0129  -122.8763      Washington         Olympia   
47   38.349   -81.6306   38.349   -81.6306   West Virginia      Charleston   
48  43.0696   -89.4239  43.0696   -89.4239       Wisconsin         Madison   
49  41.1437  -104.7962  41.1437  -104.7962         Wyoming        Cheyenne   

   abbrev  
0      AL  
1    

TypeError: the JSON object must be str, bytes or bytearray, not dict

In [79]:

import pandas as pd
import json
import plotly.express as px
import requests

info = []

for i in range(50):
  call = ("https://api.zippopotam.us/us/{0}/{1}".format(capitals['abbrev'][i],capitals['capital'][i].replace(" ","")))
  info.append(requests.get(call).text)

info

['{"country abbreviation": "US", "places": [{"place name": "Montgomery", "longitude": "-86.2578", "post code": "36101", "latitude": "32.357"}, {"place name": "Montgomery", "longitude": "-86.2085", "post code": "36102", "latitude": "32.2334"}, {"place name": "Montgomery", "longitude": "-86.2873", "post code": "36103", "latitude": "32.3462"}, {"place name": "Montgomery", "longitude": "-86.3081", "post code": "36104", "latitude": "32.373"}, {"place name": "Montgomery", "longitude": "-86.3104", "post code": "36105", "latitude": "32.3257"}, {"place name": "Montgomery", "longitude": "-86.2673", "post code": "36106", "latitude": "32.3543"}, {"place name": "Montgomery", "longitude": "-86.2799", "post code": "36107", "latitude": "32.3804"}, {"place name": "Montgomery", "longitude": "-86.3529", "post code": "36108", "latitude": "32.3417"}, {"place name": "Montgomery", "longitude": "-86.2434", "post code": "36109", "latitude": "32.3834"}, {"place name": "Montgomery", "longitude": "-86.275", "post

In [ ]:
#info = [json.loads(i) for i in info]
lat = [0]*50
lon = [0]*50

TypeError: the JSON object must be str, bytes or bytearray, not dict